### PR TITLE
Feature/relative stability metric

### PR DESCRIPTION
**Pull request by aaarrti**
- This PR implements new metrics requested in this [#116](https://github.com/understandable-machine-intelligence-lab/Quantus/issues/116) issue

### Implemented changes
 - [x] Added `RelativeInputStability`, `RelativeOutputStability`, `.RelativeRepresentationStability`
 - [x] Updated `Tutorial_ImageNet_Example_All_Metrics.ipynb` and `Tutorial_Getting_Started_with_Tensorflow.ipynb` `with these metrics
 - [x] Added test cases for new metrics
 - [x] Generated, `.rst` files for newly added classes.
 - [x] Updated README
 - [x] Added new method `get_hidden_representations` to `quantus.helpers.models.ModelInterface`, 
  as it is required for  `RelativeRepresentationStability`. Obviously, also added its implementation in TensorFlow and PyTorch.

### Minimum acceptance criteria
- ???

### CODE REVIEW

File: quantus/metrics/robustness/relative_input_stability.py [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+            Inferred from the input shape if None.
+        kwargs:
+            not used, deprecated
+        Returns
+        -------
+        relative input stability: float, np.ndarray
+            float in case `return_aggregate=True`, otherwise np.ndarray of floats
+
+
+        For each image `x`:
```
Comment (annahedstroem) [inline quantus/metrics/robustness/relative_input_stability.py:173]: This should go over parameter description

File: quantus/metrics/robustness/relative_input_stability.py [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+            4D tensor with pre-computed explanations for the x_batch.
+        device: str, optional
+            Device on which torch should perform computations.
+        softmax: boolean, optional
+            Indicates whether to use softmax probabilities or logits in model prediction.
+            This is used for this __call__ only and won't be saved as attribute. If None, self.softmax is used.
+        channel_first: boolean, optional
+            Indicates of the image dimensions are channel first, or channel last.
+            Inferred from the input shape if None.
+        kwargs:
```
Comment (annahedstroem) [inline quantus/metrics/robustness/relative_input_stability.py:165]: batch size missed, add all metrics

File: quantus/metrics/robustness/relative_representation_stability.py [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+        perturb_func_kwargs: Optional[Dict[str, ...]] = None,
+        return_aggregate: bool = False,
+        aggregate_func: Optional[Callable[[np.ndarray], np.ndarray]] = np.mean,
+        disable_warnings: bool = False,
+        display_progressbar: bool = False,
+        eps_min: float = 1e-6,
+        default_plot_func: Optional[Callable] = None,
+        layer_names: Optional[List[str]] = None,
+        layer_indices: Optional[List[int]] = None,
+        return_nan_when_prediction_changes: bool = True,
```
Comment (annahedstroem) [inline quantus/metrics/robustness/relative_representation_stability.py:57]: Should be False like the other Robustness metrics!

Comment (annahedstroem) [reply quantus/metrics/robustness/relative_representation_stability.py:57]: unless it is clearly stated in the original publication

File: quantus/metrics/robustness/relative_representation_stability.py [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+                    "function used to generate perturbations 'perturb_func' and parameters passed to it 'perturb_func_kwargs'"
+                    "number of times perturbations are sampled 'nr_samples'"
+                    "choice which internal representations to use 'layer_names', 'layer_indices'"
+                ),
+                citation='Chirag Agarwal, et. al., 2022. "Rethinking stability for attribution based explanations." https://arxiv.org/pdf/2203.06877.pdf',
+            )
+
+    def __call__(  # type: ignore
+        self,
+        model: tf.keras.Model | torch.nn.Module,
```
Comment (annahedstroem) [inline quantus/metrics/robustness/relative_representation_stability.py:144]: Very nice

File: quantus/metrics/robustness/relative_representation_stability.py [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+# Quantus project URL: <https://github.com/understandable-machine-intelligence-lab/Quantus>.
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Iterable
+import numpy as np
+from functools import partial
+
+if TYPE_CHECKING:
+    from typing import Optional, Callable, Dict, List
```
Comment (annahedstroem) [inline quantus/metrics/robustness/relative_representation_stability.py:14]: The user might not have tf or torch installed with Quantus, wouldn't it fail for them then to import it in this way?

See explanation_func for current workaround

File: quantus/metrics/robustness/relative_representation_stability.py [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+            Inferred from the input shape if None.
+        kwargs:
+            not used, deprecated
+        Returns
+        -------
+        relative representation stability: float, np.ndarray
+            float in case `return_aggregate=True`, otherwise np.ndarray of floats
+
+
+        For each image `x`:
```
Comment (annahedstroem) [inline quantus/metrics/robustness/relative_representation_stability.py:190]: Move up

File: quantus/metrics/robustness/relative_representation_stability.py [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+        kwargs:
+            Unused.
+
+        Returns
+        -------
+        ris: np.ndarray
+            The batched evaluation results.
+
+        """
+        batch_size = x_batch.shape[0]
```
Comment (annahedstroem) [inline quantus/metrics/robustness/relative_representation_stability.py:282]: While the code is quite self-explanatory, for a new user, it might not be. Please look at any other metrics and add similar comments in the three evaluate_batch implementations for this PR

File: quantus/metrics/robustness/relative_output_stability.py [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+        )
+
+        if a_batch is None:
+            a_batch = self.generate_normalised_explanations_batch(
+                x_batch, y_batch, _explain_func
+            )
+        logits = model.predict(x_batch)
+        ros_batch = np.zeros(shape=[self._nr_samples, x_batch.shape[0]])
+
+        for index in range(self._nr_samples):
```
Comment (annahedstroem) [inline quantus/metrics/robustness/relative_output_stability.py:281]: Great just see my comment about commenting!

File: quantus/metrics/robustness/relative_output_stability.py [hunk 9]
```diff
@@ -0,0 +1,314 @@
...
+            explain_func_kwargs=explain_func_kwargs,
+            a_batch=a_batch,
+            device=device,
+            softmax=softmax,
+            channel_first=channel_first,
+            model_predict_kwargs=model_predict_kwargs,
+            s_batch=None,
+            batch_size=batch_size,
+        )
+        return list(result) if isinstance(result, Iterable) else float(result)  # noqa
```
Comment (annahedstroem) [inline quantus/metrics/robustness/relative_output_stability.py:198]: for readability, make it several lines (matter or taste but still)

File: quantus/metrics/robustness/relative_output_stability.py [hunk 10]
```diff
@@ -0,0 +1,314 @@
...
+        default_plot_func: callable
+            Callable that plots the metrics result.
+        eps_min: float
+            Small constant to prevent division by 0 in relative_stability_objective, default 1e-6.
+        return_nan_when_prediction_changes: boolean
+            When set to true, the metric will be evaluated to NaN if the prediction changes after the perturbation is applied, default=True.
+        """
+
+        if normalise_func is None:
+            normalise_func = normalise_by_negative
```
Comment (annahedstroem) [inline quantus/metrics/robustness/relative_output_stability.py:93]: We have updated the default normalise function to `normalise_by_second_moment_estimate` please update in the 3 metrics

File: quantus/metrics/robustness/relative_input_stability.py [hunk 11]
```diff
@@ -0,0 +1,309 @@
...
+        denominator += (denominator == 0) * self._eps_min
+        return nominator / denominator
+
+    def evaluate_batch(
+        self,
+        model: ModelInterface,
+        x_batch: np.ndarray,
+        y_batch: np.ndarray,
+        a_batch: Optional[np.ndarray],
+        *args,
```
Comment (annahedstroem) [inline quantus/metrics/robustness/relative_input_stability.py:238]: Do we really need these? args and kwargs in signature

**Review by annahedstroem — COMMENTED** [review]
Well done. A few comments here and there. Ping me when its done and ready for a second review!

File: quantus/metrics/robustness/relative_input_stability.py (hunk 11 above)
Comment (aaarrti) [reply quantus/metrics/robustness/relative_input_stability.py:238]: Not really, but without `*args`, PyCharm complains that "Signature of method 'RelativeInputStability.evaluate_batch()' does not match signature of the base method in class 'BatchedPerturbationMetric'".

File: quantus/metrics/robustness/relative_output_stability.py (hunk 10 above)
Comment (aaarrti) [reply quantus/metrics/robustness/relative_output_stability.py:93]: I was unable to find a function named `normalise_by_second_moment_estimate`. I think this could have been a typo, and you meant `normalise_by_average_second_moment_estimate`. Moreover, this function is not used in any of the metrics as default 
image

File: quantus/metrics/robustness/relative_input_stability.py (hunk 11 above)
Comment (aaarrti) [reply quantus/metrics/robustness/relative_input_stability.py]: Removed `*args`.

File: quantus/metrics/robustness/relative_output_stability.py (hunk 10 above)
Comment (aaarrti) [reply quantus/metrics/robustness/relative_output_stability.py]: changed to 
```       
 if normalise_func is None:
      normalise_func = normalise_by_average_second_moment_estimate
```

File: quantus/metrics/robustness/relative_output_stability.py (hunk 9 above)
Comment (aaarrti) [reply quantus/metrics/robustness/relative_output_stability.py]: Done for all 3 metrics.

File: quantus/metrics/robustness/relative_representation_stability.py (hunk 7 above)
Comment (aaarrti) [reply quantus/metrics/robustness/relative_representation_stability.py:284]: done

File: quantus/metrics/robustness/relative_output_stability.py (hunk 8 above)
Comment (aaarrti) [reply quantus/metrics/robustness/relative_output_stability.py:285]: done

File: quantus/metrics/robustness/relative_representation_stability.py (hunk 6 above)
Comment (aaarrti) [reply quantus/metrics/robustness/relative_representation_stability.py]: done

File: quantus/metrics/robustness/relative_output_stability.py (hunk 10 above)
Comment (annahedstroem) [reply quantus/metrics/robustness/relative_output_stability.py]: Thank you sorry I missed the "average"

**Review by annahedstroem — COMMENTED** [review]
OK. Ready for merge.
